<a id="TableOfContents"></a>
# Table of Contents:
<li><a href='#imports'>Imports</a></li>
<li><a href="#Q1">Question 1</a></li>
<li><a href="#Q2">Question 2</a></li>
<li><a href="#Q3">Question 3</a></li>
<li><a href="#Q4">Question 4</a></li>
<li><a href="#Q5">Question 5</a></li>
<li><a href="#Bonus1">Bonus-1</a></li>
<li><a href="#Bonus2">Bonus-2</a></li>
<li><a href="#Bonus3">Bonus-3</a></li>
<li><a href="#BonusBonus">Bonus-Bonus</a></li>

<a id='imports'></a>
### Imports:
<li><a href='#TableOfContents'>Table of Contents</a></li>

In [41]:
# Basic Sheiza
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Stats
from scipy import stats

# Modeling
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

# .py file
import acquire
import prepare
import explore
import evaluation

# itertools
import itertools

In these exercises, we'll continue working with the titanic dataset and building logistic regression models. Throughout this exercise, be sure you are training, evaluation, and comparing models on the train and validate datasets. The test dataset should only be used for your final model.

In [22]:
# Get titanic dataset
titanic = prepare.prep_titanic()
titanic.sample()

survived  pclass   sex  age   fare  embark_town  alone  sex_female  \
549         1       2  male  8.0  36.75  Southampton      0           0   

     sex_male  embark_town_Cherbourg  embark_town_Queenstown  \
549         1                      0                       0   

     embark_town_Southampton  
549                        1

In [44]:
# Identify x_cols
x_cols = [col for col in titanic if col not in ['survived', 'sex', 'embark_town']]
x_cols

['pclass',
 'age',
 'fare',
 'alone',
 'sex_female',
 'sex_male',
 'embark_town_Cherbourg',
 'embark_town_Queenstown',
 'embark_town_Southampton']

In [45]:
# Identify y_cols
y_cols = 'survived'
y_cols

'survived'

In [46]:
# Split yo' data
train, validate, test = prepare.split(titanic, 'survived')

train.shape:(498, 12)
validate.shape:(214, 12)
test.shape:(179, 12)


In [47]:
# Verify yo' split
train.shape, validate.shape, test.shape

((498, 12), (214, 12), (179, 12))

In [48]:
# Identify yo' x/y train, validate, test
x_train = train[x_cols]
y_train = train[y_cols]
x_validate = validate[x_cols]
y_validate = validate[y_cols]
x_test = test[x_cols]
y_test = test[y_cols]

In [49]:
# Create train models df
# Create baseline column
train_models = pd.DataFrame(train.survived)
train_models['baseline'] = 0
train_models

survived  baseline
474         0         0
370         1         0
573         1         0
110         0         0
167         0         0
..        ...       ...
735         0         0
163         0         0
770         0         0
196         0         0
94          0         0

[498 rows x 2 columns]

<a id='Q1'></a>
### 1. Create a model that includes only age, fare, and pclass. Does this model perform better than your baseline?
<li><a href='#TableOfContents'>Table of Contents</a></li>

In [52]:
x_q1 = ['age', 'fare', 'pclass']
x_q1_train = train[x_q1]

In [53]:
lr1 = LogisticRegression()
lr1.fit(x_q1_train, y_train)
model1 = lr1.predict(x_q1_train)
train_models['model1'] = model1

In [54]:
evaluation.accuracy(train_models, 'survived')

survived: 100.00%

baseline: 61.65%

model1: 74.50%

HIGHEST VALUE = model1: 74.50%
LOWEST VALUE = baseline: 61.65%


Model-1 performs better than the baseline...

<a id='Q2'></a>
### 2. Include sex in your model as well. Note that you'll need to encode or create a dummy variable of this feature before including it in a model.
<li><a href='#TableOfContents'>Table of Contents</a></li>

In [55]:
x_cols

['pclass',
 'age',
 'fare',
 'alone',
 'sex_female',
 'sex_male',
 'embark_town_Cherbourg',
 'embark_town_Queenstown',
 'embark_town_Southampton']

In [56]:
x_q2 = ['age', 'fare', 'pclass', 'sex_female', 'sex_male']
x_q2_train = train[x_q2]

In [57]:
lr2 = LogisticRegression()
lr2.fit(x_q2_train, y_train)
model2 = lr2.predict(x_q2_train)
train_models['model2'] = model2

In [58]:
evaluation.accuracy(train_models, 'survived')

survived: 100.00%

baseline: 61.65%

model1: 74.50%

model2: 79.32%

HIGHEST VALUE = model2: 79.32%
LOWEST VALUE = baseline: 61.65%


Including 'sex' columns increased accuracy...

<a id='Q3'></a>
### 3. Try out other combinations of features and models.
<li><a href='#TableOfContents'>Table of Contents</a></li>

In [89]:
combos = []
for i in range(2, len(x_cols) + 1):
    combos += itertools.combinations(x_cols, i)
combolist = [list(combo) for combo in combos]

In [90]:
modelnum = 2
for x in combolist:
    modelnum += 1
    lr = LogisticRegression()
    lr.fit(train[x], y_train)
    model = lr.predict(train[x])
    train_models[modelnum] = model

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: Perfor

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor p

/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: Perfor

/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: 

/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/var/folders/6x/kctcvd8x6l18t57t92kmqy6w0000gn/T/ipykernel_15117/3039067025.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  train_models[modelnum] = model
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: 

In [99]:
def test():
    traindict = {
        'model': ['survived', 'baseline', '1', '2'],
        'features': ['None', 'None', ['age', 'fare', 'pclass'], ['age', 'fare', 'pclass', 'sex_male', 'sex_female']],
        'train_score': [100, 0, 0, 0],
        'validate_score': [100, 0, 0, 0]
    }
    modelnum = 2
    for x in combolist:
        modelnum += 1
        lr = LogisticRegression()
        lr.fit(train[x], y_train)
        model = lr.predict(train[x])
        traindict['model'].append(modelnum)
        traindict['features'].append(x)
        traindict['train_score'].append(lr.score(train[x], y_train))
        traindict['validate_score'].append(lr.score(validate[x], y_validate)) 
    return traindict

In [100]:
newdict = test()

/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regressi

/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regressi

/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regressi

In [102]:
newdf = pd.DataFrame(newdict)
newdf

model                                           features  train_score  \
0    survived                                               None   100.000000   
1    baseline                                               None     0.000000   
2           1                                [age, fare, pclass]     0.000000   
3           2          [age, fare, pclass, sex_male, sex_female]     0.000000   
4           3                                      [pclass, age]     0.728916   
..        ...                                                ...          ...   
501       500  [pclass, age, fare, sex_female, sex_male, emba...     0.795181   
502       501  [pclass, age, alone, sex_female, sex_male, emb...     0.805221   
503       502  [pclass, fare, alone, sex_female, sex_male, em...     0.803213   
504       503  [age, fare, alone, sex_female, sex_male, embar...     0.787149   
505       504  [pclass, age, fare, alone, sex_female, sex_mal...     0.801205   

     validate_score  
0        100.000000  
1          0.000000  
2          0.000000  
3          0.000000  
4          0.640187  
..              ...  
501        0.742991  
502        0.738318  
503        0.724299  
504        0.738318  
505        0.742991  

[506 rows x 4 columns]

In [92]:
evaluation.accuracy(train_models, 'survived')

survived: 100.00%

baseline: 61.65%

model1: 74.50%

model2: 79.32%

3: 72.89%

4: 70.48%

5: 74.30%

6: 79.32%

7: 79.32%

8: 70.48%

9: 70.28%

10: 70.88%

11: 67.27%

12: 66.47%

13: 79.32%

14: 79.32%

15: 62.85%

16: 61.65%

17: 62.25%

18: 69.08%

19: 78.51%

20: 78.51%

21: 68.88%

22: 67.87%

23: 68.88%

24: 79.32%

25: 79.32%

26: 66.47%

27: 66.47%

28: 66.47%

29: 79.32%

30: 79.32%

31: 79.32%

32: 79.32%

33: 79.32%

34: 79.32%

35: 79.32%

36: 63.05%

37: 63.05%

38: 63.05%

39: 74.50%

40: 74.50%

41: 78.92%

42: 78.92%

43: 73.29%

44: 73.29%

45: 73.09%

46: 74.30%

47: 78.51%

48: 78.51%

49: 70.48%

50: 70.28%

51: 70.88%

52: 79.52%

53: 79.52%

54: 74.30%

55: 73.69%

56: 74.30%

57: 79.32%

58: 78.51%

59: 79.32%

60: 78.51%

61: 78.51%

62: 79.32%

63: 78.51%

64: 70.28%

65: 70.28%

66: 70.28%

67: 70.08%

68: 78.51%

69: 78.51%

70: 68.67%

71: 67.27%

72: 68.27%

73: 79.32%

74: 79.32%

75: 66.27%

76: 66.47%

77: 66.06%

78: 79.32%

79: 79.32%

80: 79.32%

81

In [93]:
train_models

survived  baseline  model1  model2  3  4  5  6  7  8  ...  495  496  497  \
474         0         0       0       1  0  0  0  1  1  0  ...    1    1    1   
370         1         0       1       1  1  1  1  0  0  1  ...    1    1    1   
573         1         0       0       1  0  0  0  1  1  0  ...    1    1    1   
110         0         0       1       0  1  1  1  0  0  1  ...    0    0    0   
167         0         0       0       0  0  0  0  1  1  0  ...    0    0    0   
..        ...       ...     ...     ... .. .. .. .. .. ..  ...  ...  ...  ...   
735         0         0       0       0  0  0  0  0  0  0  ...    0    0    0   
163         0         0       0       0  0  0  0  0  0  0  ...    0    0    0   
770         0         0       0       0  0  0  0  0  0  0  ...    0    0    0   
196         0         0       0       0  0  0  0  0  0  0  ...    0    0    0   
94          0         0       0       0  0  0  0  0  0  0  ...    0    0    0   

     498  499  500  501  502  503  504  
474    1    1    1    1    0    1    1  
370    1    1    1    1    1    0    1  
573    1    1    1    1    1    1    1  
110    0    0    0    0    0    0    0  
167    0    0    0    0    1    1    0  
..   ...  ...  ...  ...  ...  ...  ...  
735    0    0    0    0    0    0    0  
163    0    0    0    0    0    0    0  
770    0    0    0    0    0    0    0  
196    0    0    0    0    0    0    0  
94     0    0    0    0    0    0    0  

[498 rows x 506 columns]

<a id='Q4'></a>
### 4. Use you best 3 models to predict and evaluate on your validate sample.
<li><a href='#TableOfContents'>Table of Contents</a></li>

In [108]:
newdf.sort_values(by='train_score', ascending=False).head(4)

model                                           features  train_score  \
0    survived                                               None   100.000000   
479       478  [pclass, age, alone, sex_male, embark_town_Che...     0.809237   
399       398  [pclass, age, alone, sex_female, embark_town_C...     0.809237   
401       400  [pclass, age, alone, sex_female, embark_town_Q...     0.809237   

     validate_score  
0        100.000000  
479        0.738318  
399        0.733645  
401        0.738318

Best 3 Training Models:
- 479:
- 399:
- 401:

<a id='Q5'></a>
### 5. Choose you best model from the validation performation, and evaluate it on the test dataset. How do the performance metrics compare to validate? to train?
<li><a href='#TableOfContents'>Table of Contents</a></li>

In [109]:
newdf.sort_values(by='validate_score', ascending=False).head(2)

model                                           features  train_score  \
0    survived                                               None   100.000000   
496       495  [pclass, age, fare, alone, sex_female, sex_mal...     0.801205   

     validate_score  
0        100.000000  
496        0.752336

Best Validate Model:
- 496

<a id='Bonus1'></a>
### Bonus-1. How do different strategies for handling the missing values in the age column affect model performance?
<li><a href='#TableOfContents'>Table of Contents</a></li>

<a id='Bonus2'></a>
### Bonus-2. How do different strategies for encoding sex affect model performance?
<li><a href='#TableOfContents'>Table of Contents</a></li>

<a id='Bonus3'></a>
### Bonus-3. scikit-learn's LogisticRegression classifier is actually applying <a href='https://en.wikipedia.org/wiki/Regularized_least_squares'>a regularization penalty to the coefficients</a> by default. This penalty causes the magnitude of the coefficients in the resulting model to be smaller than they otherwise would be. This value can be modified with the C hyper parameter. Small values of C correspond to a larger penalty, and large values of C correspond to a smaller penalty.
<li><a href='#TableOfContents'>Table of Contents</a></li>

Try out the following values for C and note how the coefficients and the model's performance on both the dataset it was trained on and on the validate split are affected.

C = .01, .1, 1, 10, 100, 1000

<a id='BonusBonus'></a>
### Bonus-Bonus. how does scaling the data interact with your choice of C?
<li><a href='#TableOfContents'>Table of Contents</a></li>